ecl/hql/hqlgram2.cpp

```diff
@@ -8992,6 +8992,34 @@ bool HqlGram::convertAllToAttribute(attribute &atr)
     return true;
 }
 
+void HqlGram::setPluggableModeExpr(attribute & targetAttr, attribute & pluginAttr, attribute & options)
+{
+    IAtom * fileFormatAtom = lower(pluginAttr.getId());
+    StringBuffer fileFormat(fileFormatAtom->queryStr());
+    const char * fileFormatStr = fileFormat.str();
+    bool isSupportedFileType = false;
+
+    // TODO: Hardcoded tests for known filetype plugins; should be
+    // an iteration through a list of available types, checking for names
+    isSupportedFileType = strsame(fileFormatStr, "parquet");
+
+    if (isSupportedFileType)
+    {
+        // Do we just cite the name of the plugin, or perhaps something
+        // like a unique ID from the plugin's factory (assuming there is one)?
+        OwnedHqlExpr fileTypeExp = createExprAttribute(fileTypeAtom, createConstant(fileFormatStr));
+        HqlExprArray args;
+        options.unwindCommaList(args);
+        // Make sure the filetype is first in the args list
+        args.add(*LINK(fileTypeExp), 0);
```

Comment:
There are two options here.  If the type is always first you don't need to wrap it in an attribute.  

```
        OwnedHqlExpr fileTypeExp = createConstant(fileFormatStr);
        HqlExprArray args;
        // Make sure the filetype is first in the args list
        args.append(*LINK(fileTypeExpr));
        options.unwindCommaList(args);
        targetAttr.setExpr(createValue(no_filetype, makeNullType(), args));

``` 
(NOTE: unwindCommaList will append to the array.)

or even use an attribute instead of a string:

```
        HqlExprArray args;
        // Make sure the filetype is first in the args list
        args.append(*createExprAttribute(fileFormatAtom));
        options.unwindCommaList(args);
        targetAttr.setExpr(createValue(no_filetype, makeNullType(), args));
``` 

If you are adding it as an attribute then it doesn't need to go first (although there is no harm if it does).
I would probably opt for 1 of the first 2 items.

Reply:
I will need to extract the file type name (e.g. "parquet") from the file format options list later on, during code generation.  I am currently doing that by searching for a fileTypeAtom.  If I don't wrap the plugin name in an attribute, how would I know to skip that option for only plugin-provided file types?  (I am hoping this makes sense.)